code/data_structures/src/list/singly_linked_list/operations/find/search_element_python.py

```diff
@@ -0,0 +1,73 @@
+class Node:
+    '''
+    A Node in a singly-linked list
+
+    Parameters
+    -------------------
+    data :
+        The data to be stored in the node
+    next:
+        The link to the next node in the singly-linked list
+    '''
+
+    def __init__(self, data=None, next=None):
+        self.data = data
+        self.next = next
+
+    def __repr__(self):
+        ''' Node representation as required'''
+        return (self.data)
+
+
+class SinglyLinkedList:
+    def __init__(self):
+        self.head = None
+
+    def append(self, data):
+        '''
+        Inserts New data at the ending of the Linked List
+        Takes O(n) time
+        '''
+        new_node = Node(data=data)
+        if not self.head:
```

Comment:
`if self.head is None:`

Reply:
In addition, you could add an adapter class inherit from `class Node` and `return True` from the magic function `def __bool__:` and then test that append new element to `self.head` if it is None.

```
class Node:
    '''
    A Node in a singly-linked list
    Parameters
    -------------------
    data :
        The data to be stored in the node
    next:
        The link to the next node in the singly-linked list
    '''

    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next

    def __repr__(self):
        ''' Node representation as required'''
        return (self.data)


class NodeAdapter1(Node):
    def __bool__(self):
        return False


class NodeAdapter2(Node):
    def __bool__(self):
        return True


class SinglyLinkedList:
    def __init__(self, NodeType):
        self.head = None
        self._NodeType = NodeType

    def append(self, data):
        '''
        Inserts New data at the ending of the Linked List
        Takes O(n) time
        '''
        new_node = self._NodeType(data=data)
        if not self.head:
            self.head = new_node
            return

        current_node = self.head
        while current_node.next is not None:
            current_node = current_node.next

        current_node.next = new_node

    def find(self, data):
        '''
        Returns the first index at which the element is found
        Returns -1 if not found
        Takes O(n) time
        '''

        if self.head is not None:
            current_node = self.head
            index = 0
            while current_node is not None:
                if current_node.data == data:
                    return index
                current_node = current_node.next
                index += 1
        return -1

    def __repr__(self):
        '''
        Gives a string representation of the list in the
        given format:
        a -> b -> c -> d
        '''
        all_nodes = []
        current_node = self.head
        while current_node:
            all_nodes.append(str(current_node.data))
            current_node = current_node.next
        if len(all_nodes) > 0:
            return ' -> '.join(all_nodes)
        else:
            return "Empty"


class SinglyLinkedList2:
    def __init__(self, NodeType):
        self.head = None
        self._NodeType = NodeType

    def append(self, data):
        '''
        Inserts New data at the ending of the Linked List
        Takes O(n) time
        '''
        new_node = self._NodeType(data=data)
        if self.head is None:
            self.head = new_node
            return

        current_node = self.head
        while current_node.next is not None:
            current_node = current_node.next

        current_node.next = new_node

    def find(self, data):
        '''
        Returns the first index at which the element is found
        Returns -1 if not found
        Takes O(n) time
        '''

        if self.head is not None:
            current_node = self.head
            index = 0
            while current_node is not None:
                if current_node.data == data:
                    return index
                current_node = current_node.next
                index += 1
        return -1

    def __repr__(self):
        '''
        Gives a string representation of the list in the
        given format:
        a -> b -> c -> d
        '''
        all_nodes = []
        current_node = self.head
        while current_node is not None:
            all_nodes.append(str(current_node.data))
            current_node = current_node.next
        if len(all_nodes) > 0:
            return ' -> '.join(all_nodes)
        else:
            return "Empty"


def main():
    '''
    sXY: X is list class version, Y is Node class version,
    SinglyLinkedList has some bug,
    that is fixed in SinglyLinkedList2.
    '''
    s10 = SinglyLinkedList(Node)
    s10.append(1)
    s10.append(2)
    s10.append(3)
    s11 = SinglyLinkedList(NodeAdapter1)
    s11.append(1)
    s11.append(2)
    s11.append(3)
    s12 = SinglyLinkedList(NodeAdapter2)
    s12.append(1)
    s12.append(2)
    s12.append(3)

    s20 = SinglyLinkedList2(Node)
    s20.append(1)
    s20.append(2)
    s20.append(3)
    s21 = SinglyLinkedList2(NodeAdapter1)
    s21.append(1)
    s21.append(2)
    s21.append(3)
    s22 = SinglyLinkedList2(NodeAdapter2)
    s22.append(1)
    s22.append(2)
    s22.append(3)

    if str(s10) != str(s11) != str(s12):
        print("s1 has wrong value")
    if str(s20) != str(s21) != str(s22):
        print("s2 has wrong value")

if __name__ == '__main__':
    main()
```